### PR TITLE
Execute autoscaler synchronously on scheduler thread

### PR DESCRIPTION
**Pull request by corindwyer**
- change autoscaler to be synchronously executed as part of the scheduling iteration while keeping the callback execution on a separate thread
- change shortfall analysis implementation to be synchronously executed as part of the autoscaler
- add withAutoscaleDisabledVmDurationInSecs for users that want to disable VMs for a fixed duration that is different from the cooldown value

### CODE REVIEW

File: fenzo-core/src/main/java/com/netflix/fenzo/StateMonitor.java [hunk 1]
```diff
@@ -16,27 +16,21 @@
...
             throw new IllegalStateException();
-        return new AutoCloseable() {
-            @Override
-            public void close() throws Exception {
-                if(!lock.compareAndSet(true, false))
-                    throw new IllegalStateException();
-            }
+        }
+        return () -> {
+            if (!lock.tryLock())
```
Comment (tbak) [inline fenzo-core/src/main/java/com/netflix/fenzo/StateMonitor.java]: Release lock?

File: fenzo-core/src/main/java/com/netflix/fenzo/TaskScheduler.java [hunk 2]
```diff
@@ -377,6 +378,23 @@ public Builder withDelayAutoscaleDownBySecs(long delayAutoscaleDownBySecs) {
...
+         * value used will be the {@link AutoScaleRule#getCoolDownSecs()} value. If the supplied {@link AutoScaleAction}
+         * does not actually terminate the instance in this time frame then VM will become enabled.
+
+         * @param disabledVmDurationInSecs Disable VMs about to be terminated by this many seconds.
+         * @return this same {@code Builder}, suitable for further chaining or to build the {@link TaskScheduler}
+         * @throws IllegalArgumentException if {@code disabledVmDurationInSecs} is not greater than 0.
+         * @see <a href="https://github.com/Netflix/Fenzo/wiki/Autoscaling">Autoscaling</a>
+         */
+        public Builder withAutoscaleDisabledVmDurationInSecs(long disabledVmDurationInSecs) {
+            if(disabledVmDurationInSecs > 0L)
```
Comment (tbak) [inline fenzo-core/src/main/java/com/netflix/fenzo/TaskScheduler.java]: <= 0?

Comment (corindwyer) [inline fenzo-core/src/main/java/com/netflix/fenzo/TaskScheduler.java]: fixed

File: fenzo-core/src/main/java/com/netflix/fenzo/StateMonitor.java (hunk 1 above)
Comment (corindwyer) [inline fenzo-core/src/main/java/com/netflix/fenzo/StateMonitor.java]: Yes, we need to release the lock here. Still need to confirm if this class is responsible for preventing multithreaded calls to the scheduling loop or if it is even supposed to protect the main thread for making multiple calls.

File: fenzo-core/src/main/java/com/netflix/fenzo/InternalVMCloner.java [hunk 3]
```diff
@@ -42,38 +43,45 @@
...
-                if (values != null && values.length == 2)
-                    disk = Math.max(disk, values[0] + values[1]);
-                values = resourceStatus.get(VMResource.Network);
-                if (values != null && values.length == 2)
-                    network = Math.max(network, values[0] + values[1]);
-                values = resourceStatus.get(VMResource.Ports);
-                if (values != null && values.length == 2)
-                    ports = Math.max(ports, values[0] + values[1]);
+            for (AssignableVirtualMachine avm : avms) {
+                Map<VMResource, Double> maxResources = avm.getMaxResources();
```
Comment (spodila) [inline fenzo-core/src/main/java/com/netflix/fenzo/InternalVMCloner.java]: Need to confirm if maxResources is set already by the time this is called

File: fenzo-core/src/main/java/com/netflix/fenzo/StateMonitor.java [hunk 4]
```diff
@@ -16,27 +16,21 @@
...
 /**
  * A monitor to ensure scheduler state is not compromised by concurrent calls that are disallowed.
  */
 class StateMonitor {
-    private final AtomicBoolean lock;
-
-    StateMonitor() {
-        lock = new AtomicBoolean(false);
-    }
+    private final ReentrantLock lock = new ReentrantLock();
```
Comment (spodila) [inline fenzo-core/src/main/java/com/netflix/fenzo/StateMonitor.java]: Why is this needed to be reentrant?

File: fenzo-core/src/main/java/com/netflix/fenzo/TaskScheduler.java [hunk 5]
```diff
@@ -466,6 +484,9 @@ private TaskScheduler(Builder builder) {
                 autoScaler.setDelayScaleDownBySecs(builder.delayAutoscaleDownBySecs);
             if(builder.delayAutoscaleUpBySecs > 0L)
                 autoScaler.setDelayScaleUpBySecs(builder.delayAutoscaleUpBySecs);
+            if (builder.disabledVmDurationInSecs > 0L) {
+                autoScaler.setDisabledVmDurationInSecs(builder.disabledVmDurationInSecs);
```
Comment (spodila) [inline fenzo-core/src/main/java/com/netflix/fenzo/TaskScheduler.java]: There is a risk in this being separate from cooldown secs. For example, the VM maybe enabled too soon and start getting assignments before it is terminated.

Comment (corindwyer) [inline fenzo-core/src/main/java/com/netflix/fenzo/TaskScheduler.java]: added javadoc to outline that should value should be greater than the cooldown period and is only useful if the autoscaling action may take more time to execute than your cooldown period.

File: fenzo-core/src/main/java/com/netflix/fenzo/InternalVMCloner.java (hunk 3 above)
Comment (corindwyer) [inline fenzo-core/src/main/java/com/netflix/fenzo/InternalVMCloner.java]: we confirmed this is set because the pseudo scheduling happens after the main scheduling.
